### PR TITLE
Populate game backend

### PR DESCRIPTION
**Pull request by ShreyamsJain**
Sets the pieces in its initial x and y position.
Performs x and y position validation as well as validation for number of pieces created.

### CODE REVIEW

File: db/migrate/20160925060545_change_column_name_in_pieces.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class ChangeColumnNameInPieces < ActiveRecord::Migration
+  def change
+  	rename_column :pieces, :type, :piece_name
```
Comment (lasley) [inline db/migrate/20160925060545_change_column_name_in_pieces.rb]: What is the reasoning behind this change?

File: app/controllers/pieces_controller.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
+class PiecesController < ApplicationController
+  def new
```
Comment (lasley) [inline app/controllers/pieces_controller.rb:2]: How can you prove this works without tests?

File: app/controllers/pieces_controller.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      Piece.create(piece_name: 'pawn_white', x_position: pawn_pos, y_position: 2)
+      Piece.create(piece_name: 'pawn_black', x_position: pawn_pos, y_position: 7)
+      pawn_pos += 1
+    end
+    render json: Piece.all
+  end
+
+  private
+
+  def set_white_pieces
```
Comment (lasley) [inline app/controllers/pieces_controller.rb:11]: How can you prove this works without tests?

File: app/controllers/pieces_controller.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    white_pieces = %w(rook_white knight_white bishop_white king_white queen_white bishop_white knight_white rook_white)
+    white_x_pos = 1
+    white_pieces.each do |wp|
+      Piece.create(piece_name: wp, x_position: white_x_pos, y_position: 1)
+      white_x_pos += 1
+    end
+  end
+
+  def set_black_pieces
```
Comment (lasley) [inline app/controllers/pieces_controller.rb:22]: How can you prove this works without tests?

File: app/models/piece.rb [hunk 5]
```diff
@@ -1,4 +1,13 @@
 class Piece < ActiveRecord::Base
   belongs_to :game
   belongs_to :player
+  before_create :validate_max_pieces
+  enum piece_name: [:king_black, :king_white, :queen_black, :queen_white, :rook_black, :rook_white, :bishop_black, :bishop_white, :knight_black, :knight_white, :pawn_black, :pawn_white]
+  validates :x_position, numericality: { greater_than_or_equal_to: 1, less_than_or_equal_to: 8 }, allow_nil: false
+  validates :y_position, numericality: { greater_than_or_equal_to: 1, less_than_or_equal_to: 8 }
+
+  def validate_max_pieces
```
Comment (lasley) [inline app/models/piece.rb]: How can you prove this works without tests?

File: coverage/.last_run.json [hunk 6]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (lasley) [inline coverage/.last_run.json]: Please remove this file. It should be in the gitignore on master now too, so it should not continue to be added.

File: coverage/.resultset.json [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (lasley) [inline coverage/.resultset.json]: Please remove this file. It should be in the gitignore on master now too, so it should not continue to be added.

File: app/controllers/pieces_controller.rb [hunk 8]
```diff
@@ -0,0 +1,33 @@
+class PiecesController < ApplicationController
+  def new
+    set_white_pieces
+    set_black_pieces
+    pawn_pos = 1
+    8.times do
+      Piece.create(piece_name: 'pawn_white', x_position: pawn_pos, y_position: 2)
+      Piece.create(piece_name: 'pawn_black', x_position: pawn_pos, y_position: 7)
+      pawn_pos += 1
```
Comment (lasley) [inline app/controllers/pieces_controller.rb]: This seems unnecessary. You should be able to obtain the loop iteration count from the loop, then use it +1 for the x_pos

File: app/models/piece.rb [hunk 9]
```diff
@@ -1,4 +1,13 @@
...
+    errors[:base] << 'Only 32 pieces are allowed.' if Piece.count == 32
```
Comment (lasley) [inline app/models/piece.rb]: I feel like this will fail after another game is created. Please add a test to prove me wrong.

**Review by lasley — CHANGES_REQUESTED** [review]

File: db/migrate/20160925060545_change_column_name_in_pieces.rb (hunk 1 above)
Comment (ShreyamsJain) [reply db/migrate/20160925060545_change_column_name_in_pieces.rb]: @lasley When i was writing validations for the type column, i was getting an error," Invalid Single table inheritance type". Then i found out that 'type' is a reserved keyword in ActiveRecord. So i changed it. Later i checked @wdehaes 's code for STI branch, i found out he had created sub classes for pieces. Then i was wondering if using type as column was preplanned so it validates using only subclasses. Am i correct? I thought of removing the entire validation and renaming the column back to type, but wanted your reviews before changing it.

Comment (lasley) [reply db/migrate/20160925060545_change_column_name_in_pieces.rb]: You are correct that `type` is used for STI & that the validations are already under the hood ;)

File: coverage/.resultset.json [hunk 10]
```diff
@@ -1,7 +1,1556 @@
 {
```
Comment (lasley) [inline coverage/.resultset.json:2]: Remove this file

File: spec/factories/bishop.rb [hunk 11]
```diff
@@ -1,4 +1,7 @@
 FactoryGirl.define do
   factory :bishop do
+    player_id 1
```
Comment (lasley) [inline spec/factories/bishop.rb]: How do you know this player exists? You should use an association https://github.com/thoughtbot/factory_girl/blob/master/GETTING_STARTED.md

**Review by lasley — CHANGES_REQUESTED** [review]

File: coverage/.resultset.json (hunk 10 above)
Comment (wdehaes) [reply coverage/.resultset.json:2]: I think I managed to remove these

**Comment by wdehaes** [comment]
@ShreyamsJain I made some minor adjustments, other than that the code looks good to me, though I am not sure if the tests cover everything (are the pieces in the right place, etc). 

Just a question, it doesn't seem like you hooked up the PiecesController#new method to the GamesController, or am I wrong? @lasley not sure if this is supposed to be part of this ticket

**Review by lasley — APPROVED** [review]

**Comment by lasley** [comment]
@wdehaes - This ticket was supposed to be model only, and has spiraled out of control in terms of scope. We should not add any more logic

**Pull request merged**